src/lib/zs_matcher.rs

```diff
@@ -37,27 +37,104 @@
-pub struct ZhangShashaConfig {}
-
-impl Default for ZhangShashaConfig {
-    fn default() -> ZhangShashaConfig {
-        ZhangShashaConfig {}
-    }
+pub struct ZhangShashaConfig {
+    src: ZSTree<FromNodeId>,
+    dst: ZSTree<ToNodeId>,
+    tree_dist: Vec<Vec<f64>>,
+    forest_dist: Vec<Vec<f64>>
 }
 
 impl ZhangShashaConfig {
     /// Create a new configuration object.
-    pub fn new() -> ZhangShashaConfig {
-        Default::default()
+    pub fn new<T: Clone + PartialEq + ToString>(src: &Arena<T, FromNodeId>,
+                                                dst: &Arena<T, ToNodeId>)
+                                                -> ZhangShashaConfig {
+        ZhangShashaConfig { src: ZSTree::new(src),
+                            dst: ZSTree::new(dst),
+                            tree_dist: vec![vec![]],
+                            forest_dist: vec![vec![]] }
+    }
+
+    /// Compute the tree distance matrix. See Section 3.2 of the paper.
+    fn compute_tree_distance<T: Clone + Debug + Eq + ToString + 'static>(&mut self,
+                                                                         store: &MappingStore<T>)
+    {
+        self.tree_dist =
+            (0..self.src.node_count + 1).map(|_| Vec::with_capacity(self.dst.node_count + 1))
+                                        .collect();
+        self.forest_dist =
+            (0..self.src.node_count + 1).map(|_| Vec::with_capacity(self.dst.node_count + 1))
+                                        .collect();
+        for i in 0..self.src.node_count + 1 {
+            for _ in 0..self.dst.node_count + 1 {
+                self.tree_dist[i].push(0.0);
+            }
+        }
+        for i in 0..self.src.node_count + 1 {
+            for _ in 0..self.dst.node_count + 1 {
+                self.forest_dist[i].push(0.0);
+            }
+        }
+        debug_assert_eq!(self.forest_dist.len(), self.src.node_count + 1);
+        debug_assert_eq!(self.forest_dist[0].len(), self.dst.node_count + 1);
+        debug_assert_eq!(self.tree_dist.len(), self.src.node_count + 1);
+        debug_assert_eq!(self.tree_dist[0].len(), self.dst.node_count + 1);
+        for i in 1..self.src.key_roots.len() {
+            for j in 1..self.dst.key_roots.len() {
+                let kr_src = self.src.key_roots[i];
+                let kr_dst = self.dst.key_roots[j];
+                self.forest_dist(kr_src, kr_dst, store);
+            }
+        }
+    }
+
+    /// Compute the forest distance matrix. See Section 3.2 of the paper.
+    fn forest_dist<T: Clone + Debug + Eq + ToString + 'static>(&mut self,
+                                                               i: usize,
+                                                               j: usize,
+                                                               store: &MappingStore<T>) {
+        self.forest_dist[self.src.lld(i) - 1][self.dst.lld(j) - 1] = 0.0;
+        for di in self.src.lld(i)..(i + 1) {
```

Comment:
I don't know if the `self.src.lld(i)` call is expensive or not, but if it is, you could assign it to a variable as it's used quite often below?